server/src/test/java/org/opensearch/index/shard/RemoteIndexShardTests.java

```diff
@@ -79,6 +97,159 @@ public void testNRTReplicaWithRemoteStorePromotedAsPrimaryCommitCommit() throws
         testNRTReplicaWithRemoteStorePromotedAsPrimary(true, true);
     }
 
+    public void testCloseShardWhileGettingCheckpoint() throws Exception {
+        String indexMapping = "{ \"" + MapperService.SINGLE_MAPPING_NAME + "\": {} }";
+        try (
+            ReplicationGroup shards = createGroup(1, getIndexSettings(), indexMapping, new NRTReplicationEngineFactory(), createTempDir())
+        ) {
+            shards.startAll();
+            IndexShard primary = shards.getPrimary();
+            final IndexShard replica = shards.getReplicas().get(0);
+            primary.refresh("Test");
+
+            final SegmentReplicationSourceFactory sourceFactory = mock(SegmentReplicationSourceFactory.class);
+            final SegmentReplicationTargetService targetService = newTargetService(sourceFactory);
+
+            // Create custom replication source in order to trigger shard close operations at specific point of segment replication
+            // lifecycle
+            SegmentReplicationSource source = new TestReplicationSource() {
+                RemoteSegmentStoreDirectory remoteDirectory;
+
+                @Override
+                public void getCheckpointMetadata(
+                    long replicationId,
+                    ReplicationCheckpoint checkpoint,
+                    ActionListener<CheckpointInfoResponse> listener
+                ) {
+                    // shard is closing while fetching metadata
+                    targetService.beforeIndexShardClosed(replica.shardId, replica, Settings.EMPTY);
+                    FilterDirectory remoteStoreDirectory = (FilterDirectory) replica.remoteStore().directory();
```

Comment:
Can we extend/wrap RemoteStoreReplicationSource here instead of re-implementing this?  Same with `testBeforeIndexShardClosedWhileCopyingFiles`.  Otherwise this test is not testing remote store implementation and the test should be generic to any ReplicationSource impl.

Reply:
Thanks @mch2. This is good catch! 
I did this in order to make target service calls (e.g. beforeIndexShardClosed) referring `TestReplicationSource` which was meant for node-node communication and made sense as `PrimaryShardReplicationSource` involves transport n/w call limitations. For remote store, we can simply extend `RemoteStoreReplicationSource` source.